patroni/dcs/etcd.py

```diff
@@ -190,31 +216,51 @@ def api_execute(self, path, method, params=None, timeout=None):
             self._load_machines_cache()
         elif not self._use_proxies and time.time() - self._machines_cache_updated > self._machines_cache_ttl:
             self._refresh_machines_cache()
-            self._machines_cache_updated = time.time()
 
-        kwargs.update(self._build_request_parameters())
+        kwargs.update(self._build_request_parameters(timeout))
 
-        if timeout is not None:
-            kwargs.update({'retries': 0, 'timeout': timeout})
+        if retry:
+            machines_cache = [self._base_uri] + self._machines_cache
 
         response = False
 
-        try:
-            some_request_failed = False
-            while not response:
-                response = self._do_http_request(request_executor, method, self._base_uri + path, **kwargs)
-
-                if response is False:
-                    some_request_failed = True
-            if some_request_failed:
-                self._refresh_machines_cache()
-        except etcd.EtcdConnectionFailed as e:
-            if isinstance(e, etcd.EtcdWatchTimedOut) and self._machines_cache:
-                self._base_uri = self._next_server()
-            else:
-                self._update_machines_cache = True
-            if not response:
-                raise
+        while True:
+            try:
+                some_request_failed = False
+                while not response:
+                    response = self._do_http_request(request_executor, method, self._base_uri + path, **kwargs)
+
+                    if response is False:
+                        if not retry:
+                            raise etcd.EtcdException
+                        some_request_failed = True
+                if some_request_failed:
+                    self._refresh_machines_cache()
+                break
+            except etcd.EtcdConnectionFailed as e:
+                if isinstance(e, etcd.EtcdWatchTimedOut):
```

Comment:
Does `EtcdWatchTimedOut` always imply a etcd problem thus justifying `self._next_server()` or can there the case where the watched key simply does not change and that triggers the timeout ?

Reply:
I moved it to the `_do_http_request` and left a comment